crypto/sphinx/src/packet.rs

```diff
@@ -0,0 +1,805 @@
+use hopr_crypto_types::crypto_traits::StreamCipher;
+use hopr_crypto_types::prelude::*;
+use hopr_primitive_types::prelude::*;
+
+use std::fmt::{Debug, Formatter};
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+use typenum::Unsigned;
+
+use crate::errors::SphinxError;
+use crate::routing::{forward_header, SphinxHeaderSpec};
+use crate::surb::{ReplyOpener, SURB};
+use crate::{
+    derivation::derive_packet_tag,
+    routing::{ForwardedHeader, RoutingInfo},
+    shared_keys::{Alpha, GroupElement, SharedKeys, SharedSecret, SphinxSuite},
+};
+
+/// Holds data that are padded up to `P + 1`.
+///
+/// Data in this instance is guaranteed to be always `P + 1` bytes-long.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct PaddedPayload<const P: usize>(Box<[u8]>);
+
+impl<const P: usize> PaddedPayload<P> {
+    /// Tag used to separate padding from data
+    pub const PADDING_TAG: u8 = 0xaa;
+
+    /// Byte used to pad the data.
+    pub const PADDING: u8 = 0x00;
+
+    /// Size of the padded data.
+    pub const SIZE: usize = P + size_of_val(&Self::PADDING_TAG);
+
+    /// Creates a new instance from the given message `msg` shorter than [`PaddedPayload::SIZE`] and pads it.
+    ///
+    /// The padding consists of prepending a [`PaddedPayload::PADDING_TAG`], preceded by as many zero bytes
+    /// to fill it up to [`PaddedPayload::SIZE`]. If data is `P` bytes-long, only the padding tag is prepended.
+    ///
+    /// If the argument's length is greater or equal to [`PaddedPayload::SIZE`], [`SphinxError::PaddingError`] is returned.
+    pub fn new(msg: &[u8]) -> Result<Self, SphinxError> {
+        if msg.len() < Self::SIZE {
+            // Zeroes followed by the PADDING_TAG and then the message
+            let mut ret = vec![Self::PADDING; Self::SIZE];
+            ret[Self::SIZE - msg.len() - 1] = Self::PADDING_TAG;
+            ret[Self::SIZE - msg.len()..].copy_from_slice(msg);
```

Comment:
Easier to work with would be to have

| size of X | X | Padding |
| --- | --- | --- |
| 2B | XB | PaddingB |

Reply:
No, this is defined by the Sphinx paper and cannot be changed. It would violate some KCA assumptions,

Reply:
I will read through the paper again.